### PR TITLE
Consolidate `pip install` calls in CI

### PR DESCRIPTION
**Pull request by Eric-Arellano**
### Summary

As @jakelishman points out in https://github.com/Qiskit/qiskit-terra/pull/9756#pullrequestreview-1331217330, calling `pip install` with all our requirements in the same call allows its resolver to better consider all the constraints. This is a "safer" implementation.

### Details and comments

This tried to stop using an editable install, which Pip recommends. But I couldn't figure out within my timebox why some CI jobs were failing saying they could not find our Rust extension. So, I gave up on that.

### CODE REVIEW

File: .azure/lint-linux.yml [hunk 1]
```diff
@@ -21,10 +21,12 @@ jobs:
           python -m pip install --upgrade pip setuptools wheel virtualenv
           virtualenv test-job
           source test-job/bin/activate
-          pip install -U -r requirements.txt -r requirements-dev.txt -c constraints.txt
-          pip install -U -c constraints.txt -e .
-          pip install -U "qiskit-aer" -c constraints.txt
-          pip install -e .
```
Comment (Eric-Arellano) [inline .azure/lint-linux.yml:27]: FYI we were calling `install -e .` twice here: https://github.com/Qiskit/qiskit-terra/pull/9756

**Comment by mtreinish** [comment]
One other thing to consider is a difference between `pip install -e .` and `pip install .` (and the sdist install) is that for the rust component editable installs are built in debug mode and full installs are using release mode. So there is a tradeoff between debug test coverage (e.g. debug mode will test for overflows, etc) and compile time vs runtime. In general I think all our test jobs were running in release mode already (except for the coverage job). I don't really have a strong opinion one way or the other. But it was something I wanted to call out as a possible consideration while we're cleaning things up.

**Review by jakelishman — COMMENTED** [review]
I think I would suggest restoring the `--upgrade` components in cases where we've coalesced all the install commands into one.  The prior `-U` option is a) outside the virtual environments, so doesn't affect the `setuptools` and `pip` installed within them and b) important to do an dependency upgrade if any of the packages installed request versions of bootstrapped packages (e.g. setuptools) at _runtime_ - we probably don't want to be stuck with whatever old version happened to be put in by the venv.  At worst the `-U` is superfluous, and at best I think it's the desired semantics: a clean, up-to-date install.

Other than that, I'm fine with this.

File: .azure/docs-linux.yml [hunk 2]
```diff
@@ -9,6 +9,7 @@ jobs:
 
     variables:
       PIP_CACHE_DIR: $(Pipeline.Workspace)/.pip
+      RUST_DEBUG: 1
```
Comment (Eric-Arellano) [inline .azure/docs-linux.yml:12]: This doesn't actually matter. Tox is using editable installs. And even after my PR to turn that off for docs (to fix macOS), we'll still have Tox set RUST_DEBUG.

But, I figure it doesn't hurt to be consistent here. Avoids us accidentally losing this in the future if someone changes tox.ini without realizing the implications.

**Comment by Eric-Arellano** [comment]
I gave up on trying to get not editable installs to work. This is ready to land now.

**Comment by Eric-Arellano** [comment]
Bump

**Review by mtreinish — APPROVED** [review]
This LGTM, it should also improve throughput slightly as pip won't need to run its depsovler as often.

**Pull request merged**